### PR TITLE
feat: Add XAI SDK integration to TensorFlow models with LIT integration

### PR DESCRIPTION
**Pull request by taiseiak**
Add automatic addition of feature attribution for TensorFlow 2 models in the LIT integration on Vertex Notebooks. Detects for Vertex Notebooks by looking for the same environment variable to check for Vertex Notebooks as the LIT library does.

Fixes b/210943910 🦕

go/local-explanations-lit-xai-notebook

### CODE REVIEW

**Comment by Ark-kun** [comment]
In the future, would it be possible to add the automatic addition of feature attribution for TensorFlow 2 models imported using `Models.upload_tensorflow_model`?

**Comment by taiseiak** [comment]
Where is `Models.upload_tensorflow_model`? I cannot find it.

**Comment by Ark-kun** [comment]
Documentation: https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.Model#google_cloud_aiplatform_Model_upload_tensorflow_saved_model

Code: https://github.com/googleapis/python-aiplatform/blob/8df74a29df0adb95fff5500fcc9d7a025012ab5e/google/cloud/aiplatform/models.py#L2857

File: google/cloud/aiplatform/explain/lit.py [hunk 1]
```diff
@@ -15,11 +15,13 @@
 # limitations under the License.
 
 from typing import Dict, List, Tuple, Union
+import os
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/explain/lit.py:18]: Imports should be in lexicographic order.

File: google/cloud/aiplatform/explain/lit.py [hunk 2]
```diff
@@ -15,11 +15,13 @@
...
 
 try:
     from lit_nlp.api import dataset as lit_dataset
     from lit_nlp.api import model as lit_model
     from lit_nlp.api import types as lit_types
+    from lit_nlp.api import dtypes as lit_dtypes
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/explain/lit.py:24]: Same here.

**Review by sasha-gitg — CHANGES_REQUESTED** [review]

**Comment by taiseiak** [comment]
> Documentation: https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.Model#google_cloud_aiplatform_Model_upload_tensorflow_saved_model
> 
> Code:
> 
> https://github.com/googleapis/python-aiplatform/blob/8df74a29df0adb95fff5500fcc9d7a025012ab5e/google/cloud/aiplatform/models.py#L2857

I think this is possible, but I don't know if we should do this automatically. We charge users for using explanations on uploaded models. I think having a parameter that sets up explanation metadata or having a function that utilizes the XAI SDK to create metadata would be better. Would those suffice?

File: google/cloud/aiplatform/explain/lit.py [hunk 3]
```diff
@@ -94,39 +97,35 @@ def __init__(
...
         self._output_types = output_types
+        self._input_tensor_name = next(iter(self._kwargs_signature))
+        self._attribution_explainer = None
+        if os.environ.get("LIT_PROXY_URL"):
+            self._set_up_attribution_explainer(model, attribution_method)
+
+    @property
+    def attribution_explainer(
+        self,
+    ) -> Union["AttributionExplainer", None]:  # noqa: F821
```
[the remark this answers is not in the document]
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py]: Done.

File: google/cloud/aiplatform/explain/lit.py [hunk 4]
```diff
@@ -148,7 +156,52 @@ def input_spec(self) -> lit_types.Spec:
...
         """Return a spec describing model outputs."""
-        return self._output_types
+        output_spec_dict = dict(self._output_types)
+        if self.attribution_explainer:
+            output_spec_dict["feature_attribution"] = lit_types.FeatureSalience(
+                signed=True
+            )
+        return output_spec_dict
+
+    def _load_model(self, model: str):
```
[the remark this answers is not in the document]
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py:166]: Done.

File: google/cloud/aiplatform/explain/lit.py [hunk 5]
```diff
@@ -148,7 +156,52 @@ def input_spec(self) -> lit_types.Spec:
...
+            raise ValueError("Please use a model with only one output tensor.")
+
+    def _set_up_attribution_explainer(
+        self, model: str, attribution_method: str = "integrated_gradients"
+    ):
+        try:
+            import explainable_ai_sdk
+            from explainable_ai_sdk.metadata.tf.v2 import SavedModelMetadataBuilder
+        except ImportError:
+            return
```
[the remark this answers is not in the document]
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py:198]: Done. Is this the correct way to log? I have seen examples in the codebase.

File: google/cloud/aiplatform/explain/lit.py [hunk 6]
```diff
@@ -148,7 +156,52 @@ def input_spec(self) -> lit_types.Spec:
...
+        _, self._kwargs_signature = serving_default.structured_input_signature
+        self._output_signature = serving_default.structured_outputs
+
+        if len(self._kwargs_signature) != 1:
+            raise ValueError("Please use a model with only one input tensor.")
+
+        if len(self._output_signature) != 1:
+            raise ValueError("Please use a model with only one output tensor.")
+
+    def _set_up_attribution_explainer(
```
[the remark this answers is not in the document]
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py:186]: Done.

File: google/cloud/aiplatform/explain/lit.py (hunk 1 above)
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py:17]: Done.

File: google/cloud/aiplatform/explain/lit.py (hunk 2 above)
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py]: Done.

**Comment by Ark-kun** [comment]
> > Documentation: https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.Model#google_cloud_aiplatform_Model_upload_tensorflow_saved_model
> > Code:
> > https://github.com/googleapis/python-aiplatform/blob/8df74a29df0adb95fff5500fcc9d7a025012ab5e/google/cloud/aiplatform/models.py#L2857
> 
> I think this is possible, but I don't know if we should do this automatically. We charge users for using explanations on uploaded models. I think having a parameter that sets up explanation metadata or having a function that utilizes the XAI SDK to create metadata would be better. Would those suffice?

I have couple of questions:

1) Are users charged extra when XAI metadata is added to a model?
2) Is there an easy way for a user to add explanation metadata to a model that would fit most users?

**Comment by taiseiak** [comment]
> > > Documentation: https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform.Model#google_cloud_aiplatform_Model_upload_tensorflow_saved_model
> > > Code:
> > > https://github.com/googleapis/python-aiplatform/blob/8df74a29df0adb95fff5500fcc9d7a025012ab5e/google/cloud/aiplatform/models.py#L2857
> > 
> > 
> > I think this is possible, but I don't know if we should do this automatically. We charge users for using explanations on uploaded models. I think having a parameter that sets up explanation metadata or having a function that utilizes the XAI SDK to create metadata would be better. Would those suffice?
> 
> I have couple of questions:
> 
> 1. Are users charged extra when XAI metadata is added to a model?
> 2. Is there an easy way for a user to add explanation metadata to a model that would fit most users?

1. Currently, no. However, this will change in the future. When a user deploys a model to an endpoint with explanation metadata, they will be charged for the explainer GKE pod for that endpoint. Users are also charged extra for using the `explain` call also as it takes computation time to calculate explanations.
2.  The most easy way right now is to use the XAI SDK (https://github.com/GoogleCloudPlatform/explainable_ai_sdk). This SDK has many convenience methods for generating metadata. This is what I used to generate metadata for this PR.

File: google/cloud/aiplatform/explain/lit.py [hunk 7]
```diff
@@ -148,7 +154,63 @@ def input_spec(self) -> lit_types.Spec:
...
+        self, model: str, attribution_method: str = "integrated_gradients"
+    ):
+        """Populates the attribution explainer attribute of the class."""
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/explain/lit.py:187]: Should contain Args section.

File: google/cloud/aiplatform/explain/lit.py [hunk 8]
```diff
@@ -148,7 +154,63 @@ def input_spec(self) -> lit_types.Spec:
...
+        try:
+            import explainable_ai_sdk
+            from explainable_ai_sdk.metadata.tf.v2 import SavedModelMetadataBuilder
+        except ImportError:
+            print(
+                "Skipping explanations because the Explainable AI SDK is not installed."
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/explain/lit.py:193]: Preference for either `logging.info` or `logging.warning`

File: tests/unit/aiplatform/test_explain_lit.py [hunk 9]
```diff
@@ -15,13 +15,15 @@
 # limitations under the License.
 
 import collections
+import os
```
Comment (sasha-gitg) [inline tests/unit/aiplatform/test_explain_lit.py:18]: Try to keep these in lexicographic order as well.

**Review by sasha-gitg — CHANGES_REQUESTED** [review]

File: tests/unit/aiplatform/test_explain_lit.py (hunk 9 above)
Comment (taiseiak) [reply tests/unit/aiplatform/test_explain_lit.py:18]: Sorry, missed this!

File: google/cloud/aiplatform/explain/lit.py (hunk 8 above)
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py:193]: Chose logging.info.

File: google/cloud/aiplatform/explain/lit.py (hunk 7 above)
Comment (taiseiak) [reply google/cloud/aiplatform/explain/lit.py:187]: Added args section.

**Review by sasha-gitg — APPROVED** [review]

**Pull request merged**